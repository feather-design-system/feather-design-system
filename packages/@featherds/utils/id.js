const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
