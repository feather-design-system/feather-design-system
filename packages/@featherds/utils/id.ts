const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
