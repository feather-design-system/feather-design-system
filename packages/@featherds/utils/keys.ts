const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
