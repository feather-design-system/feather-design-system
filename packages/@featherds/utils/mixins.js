var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
