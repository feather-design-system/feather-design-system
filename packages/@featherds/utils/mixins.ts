const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
