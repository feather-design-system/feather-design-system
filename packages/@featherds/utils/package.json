{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
