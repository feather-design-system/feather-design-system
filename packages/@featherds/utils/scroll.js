var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
