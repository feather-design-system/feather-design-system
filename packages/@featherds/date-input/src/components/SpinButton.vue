<template>
  <span
    ref="spinner"
    tabindex="-1"
    role="spinbutton"
    :aria-label="label"
    :aria-valuemin="min"
    :aria-valuemax="max"
    :aria-valuenow="modelValue || 0"
    :aria-valuetext="modelValue || placeholder"
    @keydown="handleKeyDown"
    @focus="handleFocus"
    @click.stop="handleFocus"
    >{{ displayText }}</span
  >
</template>

<script>
import { KEYCODES } from "@featherds/utils/keys";
export default {
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits: ["update:modelValue", "next", "previous"],
  props: {
    min: {
      type: Number,
      required: true,
    },
    max: {
      type: Number,
      required: true,
    },
    modelValue: {
      type: Number,
    },
    label: {
      type: String,
      required: true,
    },
    placeholder: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      input: "",
    };
  },
  computed: {
    displayText() {
      const pad = (value) => {
        let paddingLength =
          this.max.toString().length - value.toString().length;
        let padding = "";
        while (paddingLength > 0) {
          padding += "0";
          paddingLength--;
        }
        return `${padding}${value}`;
      };
      if (this.modelValue !== undefined) {
        if (this.modelValue.toString().length < this.max.toString().length) {
          return pad(this.modelValue);
        }
        return this.modelValue.toString();
      }
      if (this.input) {
        return pad(this.input);
      }
      return this.placeholder || "";
    },
  },
  methods: {
    focus() {
      this.$refs.spinner.focus();
    },
    deselect() {
      if (window.getSelection) {
        window.getSelection().removeAllRanges();
      }
    },
    parseValue(str) {
      const value = parseInt(str, 10);
      if (isNaN(value) || value < this.min || value > this.max) {
        this.$emit("update:modelValue", undefined);
      } else {
        this.$emit("update:modelValue", value);
        //if they type in the string 01 then go to next
        if (this.max.toString().length === str.length) {
          this.$emit("next");
          return;
        }
      }

      if (value * 10 >= this.max) {
        this.$emit("next");
      }
    },
    handleFocus(e) {
      this.highlight(e.target);
      this.input = "";
    },
    highlight(el) {
      if (document.createRange) {
        var range = document.createRange();
        range.selectNode(el);
        window.getSelection().removeAllRanges();
        window.getSelection().addRange(range);
      }
    },
    handleKeyDown(e) {
      const isModifiedKeyPress = (_ev) => {
        return _ev.shiftKey || _ev.ctrlKey || _ev.metaKey || _ev.altKey;
      };

      if (isModifiedKeyPress(e)) {
        return;
      }
      const stop = (_ev) => {
        _ev.stopPropagation();
        _ev.preventDefault();
      };

      //handle input for numbers
      const allowed = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
      if (allowed.indexOf(e.key) > -1) {
        const newInput = this.input + e.key;
        //if a user tries to type in 6 chars into the year only take last 4 as valid
        const charLimit = this.max.toString().length;
        if (newInput.length > charLimit) {
          this.input = newInput.substring(newInput.length - charLimit);
        } else {
          this.input = newInput;
        }
        this.parseValue(this.input);
        stop(e);
        return;
      }

      //handle arrow keys
      switch (e.keyCode) {
        case KEYCODES.LEFT:
          this.$emit("previous");
          stop(e);
          break;
        case KEYCODES.RIGHT:
          this.$emit("next");
          stop(e);
          break;
        case KEYCODES.UP:
          if (this.modelValue < this.max) {
            this.$emit("update:modelValue", this.modelValue + 1);
          }
          if (this.modelValue === undefined || this.modelValue === 0) {
            this.$emit("update:modelValue", this.min);
          }
          stop(e);
          break;
        case KEYCODES.DOWN:
          if (this.modelValue > this.min) {
            this.$emit("update:modelValue", this.modelValue - 1);
          }
          if (this.modelValue === undefined || this.modelValue === 0) {
            this.$emit("update:modelValue", this.max);
          }
          stop(e);
          break;
        case KEYCODES.DELETE:
          //reset
          this.input = "";
          this.$emit("update:modelValue", undefined);
          stop(e);
          break;
        case KEYCODES.BACKSPACE:
          this.input = this.input.slice(0, -1);

          this.parseValue(this.input);
          stop(e);

          break;
      }
    },
  },
};
</script>
<style lang="scss" scoped>
span {
  cursor: pointer;
  &:focus {
    outline: 0;
  }
}
</style>
