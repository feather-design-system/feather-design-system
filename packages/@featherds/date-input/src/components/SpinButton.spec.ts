import SpinButton from "./SpinButton.vue";
import { mount } from "@vue/test-utils";
import { Code } from "@featherds/utils/keys";
import { getCalls } from "@featherds/utils/test/calls";

const getWrapper = function (options: Record<string, any>) {
  return mount(SpinButton, options);
};

const setValue = async (
  wrapper: ReturnType<typeof getWrapper>,
  str: string
) => {
  await wrapper.trigger("focus");
  const chars = str.split("");
  return chars.reduce((prev, curr) => {
    return prev.then(() => {
      return wrapper.find("span").trigger("keydown", {
        key: curr,
      });
    });
  }, Promise.resolve());
};
const getValue = (wrapper: ReturnType<typeof getWrapper>) => {
  return wrapper.find("span").element.textContent;
};

describe("SpinButton.vue", () => {
  it("should render correctly", () => {
    const wrapper = getWrapper({
      props: {
        min: 0,
        max: 10,
        modelValue: 4,
        label: "Test",
        placeholder: "dd",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render placeholder when there is no value", () => {
    const wrapper = getWrapper({
      props: {
        min: 0,
        max: 10,
        label: "Test",
        placeholder: "dd",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render disabled", () => {
    const wrapper = getWrapper({
      propsData: {
        min: 0,
        max: 10,
        label: "Test",
        placeholder: "dd",
        disabled: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should add 0s for 1000 ", async () => {
    const wrapper = getWrapper({
      props: {
        min: 0,
        max: 1000,
        modelValue: 4,
        label: "Test",
        placeholder: "dd",
      },
    });
    expect(getValue(wrapper)).toBe("0004");

    await wrapper.setProps({
      max: 1000,
      modelValue: 32,
    });
    expect(getValue(wrapper)).toBe("0032");

    await wrapper.setProps({
      max: 100,
      modelValue: 3,
    });
    expect(getValue(wrapper)).toBe("003");

    await wrapper.setProps({
      max: 10,
      modelValue: 2,
    });
    expect(getValue(wrapper)).toBe("02");
  });
  it("should emit next when a user cannot type any more values to match the current range", async () => {
    const wrapper = getWrapper({
      props: {
        min: 1,
        max: 1000,
        label: "Test",
        placeholder: "dd",
      },
    });
    await setValue(wrapper, "200");
    expect(getCalls<[unknown]>(wrapper, "next").length).toBe(1);

    await setValue(wrapper, "99");
    expect(getCalls<[unknown]>(wrapper, "next").length).toBe(1);

    await setValue(wrapper, "909");
    expect(getCalls<[unknown]>(wrapper, "next").length).toBe(2);

    await wrapper.setProps({
      max: 31,
    });
    await setValue(wrapper, "3");
    expect(getCalls<[unknown]>(wrapper, "next").length).toBe(2);
    await setValue(wrapper, "31");
    expect(getCalls<[unknown]>(wrapper, "next").length).toBe(3);
    await setValue(wrapper, "0001");
    expect(getCalls<[unknown]>(wrapper, "next").length).toBe(4);
  });
  it("should emit update models when value is input", async () => {
    const wrapper = getWrapper({
      props: {
        min: 5,
        max: 1000,
        label: "Test",
        placeholder: "dd",
      },
    });
    await setValue(wrapper, "1");
    expect(getCalls<[number]>(wrapper, "update:modelValue").length).toBe(1);
    expect(
      getCalls<[unknown]>(wrapper, "update:modelValue")[0][0]
    ).toBeUndefined();

    await setValue(wrapper, "009");
    expect(getCalls<[number]>(wrapper, "update:modelValue")[3][0]).toBe(9); //one for each char
  });
  it("should clear value when delete is pressed", async () => {
    const wrapper = getWrapper({
      propsData: {
        modelValue: 11,
        min: 5,
        max: 1000,
        label: "Test",
        placeholder: "dd",
      },
    });

    const span = wrapper.find("span");
    await span.trigger("keydown", { code: Code.DELETE });

    expect(
      getCalls<[unknown]>(wrapper, "update:modelValue")[0][0]
    ).toBeUndefined();
    await wrapper.setProps({
      modelValue: getCalls<[unknown]>(wrapper, "update:modelValue")[0][0],
    });
    expect(getValue(wrapper)).toBe("dd");
  });
  it("should clear input when no character is typed", async () => {
    const wrapper = getWrapper({
      propsData: {
        modelValue: 11,
        min: 5,
        max: 1000,
        label: "Test",
        placeholder: "dd",
      },
    });

    const span = wrapper.find("span");
    await span.trigger("keydown", { code: Code.BACKSPACE });

    expect(
      getCalls<[unknown]>(wrapper, "update:modelValue")[0][0]
    ).toBeUndefined();
    await wrapper.setProps({
      modelValue: getCalls<[unknown]>(wrapper, "update:modelValue")[0][0],
    });
    expect(getValue(wrapper)).toBe("dd");
  });
  it("should clear last character typed when backspace is pressed", async () => {
    const wrapper = getWrapper({
      propsData: {
        modelValue: 11,
        min: 5,
        max: 1000,
        label: "Test",
        placeholder: "dd",
      },
    });

    await wrapper.find("span").trigger("keydown", {
      key: "1",
    });
    await wrapper.find("span").trigger("keydown", {
      key: "1",
    });
    await wrapper.find("span").trigger("keydown", {
      key: "9",
    });

    expect(getCalls<[number]>(wrapper, "update:modelValue")[2][0]).toBe(119);

    const span = wrapper.find("span");
    await span.trigger("keydown", { code: Code.BACKSPACE });

    expect(getCalls<[number]>(wrapper, "update:modelValue")[3][0]).toBe(11);
    expect(getValue(wrapper)).toBe("0011");
  });
});
