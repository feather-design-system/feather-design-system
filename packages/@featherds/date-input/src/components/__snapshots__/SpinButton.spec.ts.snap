// Vitest Snapshot v1

exports[`SpinButton.vue > should render correctly 1`] = `
<span
  aria-disabled="false"
  aria-invalid="false"
  aria-label="Test"
  aria-valuemax="10"
  aria-valuemin="0"
  aria-valuenow="4"
  aria-valuetext="4"
  class=""
  role="spinbutton"
  tabindex="-1"
>
  04
</span>
`;

exports[`SpinButton.vue > should render disabled 1`] = `
<span
  aria-disabled="true"
  aria-invalid="false"
  aria-label="Test"
  aria-valuemax="10"
  aria-valuemin="0"
  aria-valuenow="0"
  aria-valuetext="dd"
  class="disabled"
  role="spinbutton"
  tabindex="-1"
>
  dd
</span>
`;

exports[`SpinButton.vue > should render placeholder when there is no value 1`] = `
<span
  aria-disabled="false"
  aria-invalid="false"
  aria-label="Test"
  aria-valuemax="10"
  aria-valuemin="0"
  aria-valuenow="0"
  aria-valuetext="dd"
  class=""
  role="spinbutton"
  tabindex="-1"
>
  dd
</span>
`;
